File: openclaw-plugin/tools.ts
import { textResult, fetchJson, hasRecallConfig } from './api';
import { resolveMemoryLocator, splitParentPathAndTitle, trimSlashes } from './uri';
import { formatNode, formatBootView, normalizeSearchResults, normalizeKeywordList } from './formatters';
import { markSessionRead } from './hooks';

async function applyGlossaryMutations(pluginCfg: any, nodeUuid: string, { add = [], remove = [] }: { add?: string[]; remove?: string[] } = {}) {
  const added: string[] = [];
  const removed: string[] = [];
  for (const keyword of normalizeKeywordList(add)) {
    await fetchJson(pluginCfg, "/browse/glossary", {
      method: "POST",
      body: JSON.stringify({ keyword, node_uuid: nodeUuid }),
    });
    added.push(keyword);
  }
  for (const keyword of normalizeKeywordList(remove)) {
    await fetchJson(pluginCfg, "/browse/glossary", {
      method: "DELETE",
      body: JSON.stringify({ keyword, node_uuid: nodeUuid }),
    });
    removed.push(keyword);
  }
  return { added, removed };
}

export function registerTools(api: any, pluginCfg: any) {
  api.registerTool({
    name: "lore_status",
    label: "Lore status",
    description: "Check memory backend availability and connection health.",
    parameters: { type: "object", additionalProperties: false, properties: {} },
    async execute() {
      try {
        const data = await fetchJson(pluginCfg, "/health", { method: "GET" });
        return textResult(`Lore online\n\n${JSON.stringify(data, null, 2)}`, { ok: true, health: data, baseUrl: pluginCfg.baseUrl });
      } catch (error: any) {
        return textResult(`Lore offline: ${error.message}`, { ok: false, error: error.message, baseUrl: pluginCfg.baseUrl });
      }
    },
  });

  api.registerTool({
    name: "lore_boot",
    label: "Lore boot",
    description: "Load the boot memory view that restores long-term identity and core operating context.",
    parameters: { type: "object", additionalProperties: false, properties: {} },
    async execute() {
      try {
        const data = await fetchJson(pluginCfg, "/browse/boot", { method: "GET" });
        const content = formatBootView(data);
        return textResult(content, { ok: true, content, boot: data });
      } catch (error: any) {
        return textResult(`Lore boot failed: ${error.message}`, { ok: false, error: error.message });
      }
    },
  });

  api.registerTool({
    name: "lore_get_node",
    label: "Lore get node",
    description: "Open a memory node to inspect its full content, metadata, and nearby structure.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["uri"],
      properties: {
        uri: { type: "string", description: "Full memory URI for the node you want to open, such as core://soul." },
        nav_only: { type: "boolean", description: "If true, skip expensive glossary processing." },
        __session_id: { type: "string", description: "Internal session tracking field." },
        __session_key: { type: "string", description: "Internal session tracking field." }
      }
    },
    async execute(_id: any, params: any) {
      const navOnly = params?.nav_only === true;
      const sessionId = typeof params?.__session_id === "string" && params.__session_id.trim() ? params.__session_id.trim() : "";
      const sessionKey = typeof params?.__session_key === "string" && params.__session_key.trim() ? params.__session_key.trim() : "";
      let domain = pluginCfg.defaultDomain;
      let path = "";
      try {
        ({ domain, path } = resolveMemoryLocator(params, { defaultDomain: pluginCfg.defaultDomain, pathKey: "__unused_path", allowEmptyPath: true, label: "uri" }));
        const qs = new URLSearchParams({ domain, path, nav_only: String(navOnly) });
        const data = await fetchJson(pluginCfg, `/browse/node?${qs.toString()}`, { method: "GET" });
        const node = data?.node || {};
        if (sessionId && node?.uri) {
          await markSessionRead(pluginCfg, {
            sessionId,
            sessionKey,
            uri: node.uri,
            nodeUuid: node.node_uuid,
            source: "tool:lore_get_node",
          });
        }
        return textResult(formatNode(data), { ok: true, node, children: data?.children || [] });
      } catch (error: any) {
        return textResult(`Lore get node failed: ${error.message}`, { ok: false, error: error.message, domain, path });
      }
    },
  });

  api.registerTool({
    name: "lore_search",
    label: "Lore search",
    description: "Find relevant memories by keyword or domain when you need to locate prior knowledge.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["query"],
      properties: {
        query: { type: "string" },
        domain: { type: "string", description: "Optional domain filter to narrow the search." },
        limit: { type: "integer", minimum: 1, maximum: 100 }
      }
    },
    async execute(_id: any, params: any) {
      const query = String(params?.query || "").trim();
      const safeLimit = Number.isFinite(params?.limit) ? Math.max(1, Math.min(100, params.limit)) : 10;
      try {
        let data;
        if (hasRecallConfig(pluginCfg)) {
          data = await fetchJson(pluginCfg, `/browse/search`, {
            method: "POST",
            body: JSON.stringify({
              query,
              domain: typeof params?.domain === "string" && params.domain.trim() ? params.domain.trim() : null,
              limit: safeLimit,
              hybrid: true,
            }),
          });
        } else {
          const qs = new URLSearchParams({ query });
          if (typeof params?.domain === "string" && params.domain.trim()) qs.set("domain", params.domain.trim());
          qs.set("limit", String(safeLimit));
          data = await fetchJson(pluginCfg, `/browse/search?${qs.toString()}`, { method: "GET" });
        }
        const results = normalizeSearchResults(data);
        const meta = data?.meta || null;
        const text = results.length > 0
          ? results.map((item: any, idx: number) => {
              const parts = [`${idx + 1}. ${item.uri} (priority: ${item.priority}`];
              if (typeof item?.score === "number") parts.push(`score: ${item.score.toFixed(3)}`);
              if (Array.isArray(item?.matched_on) && item.matched_on.length > 0) parts.push(`via: ${item.matched_on.join("+")}`);
              return `${parts.join(", ")})\n   ${item.snippet}`;
            }).join("\n")
          : "No matching memories found.";
        const suffix = meta?.semantic_error ? `\n\nSemantic fallback skipped: ${meta.semantic_error}` : "";
        return textResult(`${text}${suffix}`, { ok: true, results, meta });
      } catch (error: any) {
        return textResult(`Lore search failed: ${error.message}`, { ok: false, error: error.message, query });
      }
    },
  });

  api.registerTool({
    name: "lore_list_domains",
    label: "Lore list domains",
    description: "Browse the top-level memory domains available in the memory system.",
    parameters: { type: "object", additionalProperties: false, properties: {} },
    async execute() {
      try {
        const data = await fetchJson(pluginCfg, "/browse/domains", { method: "GET" });
        const text = Array.isArray(data) && data.length > 0
          ? data.map((item: any) => `- ${item.domain} (${item.root_count})`).join("\n")
          : "No domains found.";
        return textResult(text, { ok: true, domains: data });
      } catch (error: any) {
        return textResult(`Lore list domains failed: ${error.message}`, { ok: false, error: error.message });
      }
    },
  });

  api.registerTool({
    name: "lore_create_node",
    label: "Lore create node",
    description: "Create a new long-term memory node for durable facts, rules, project knowledge, or conclusions worth keeping.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["content", "priority", "glossary"],
      properties: {
        uri: { type: "string", description: "Optional final memory URI. Use this when you already know exactly where the new memory should live." },
        domain: { type: "string", description: "Target memory domain when you are not using `uri`." },
        parent_path: { type: "string", description: "Parent location inside the chosen domain." },
        content: { type: "string" },
        priority: { type: "integer", minimum: 0 },
        title: { type: "string", description: "Final path segment for the new memory." },
        disclosure: { type: "string" },
        glossary: { type: "array", items: { type: "string" } }
      }
    },
    async execute(_id: any, params: any) {
      const glossary = normalizeKeywordList(params?.glossary);
      const body: any = {
        domain: typeof params?.domain === "string" && params.domain.trim() ? params.domain.trim() : pluginCfg.defaultDomain,
        parent_path: typeof params?.parent_path === "string" ? trimSlashes(params.parent_path) : "",
        content: String(params?.content || ""),
        priority: Number(params?.priority),
      };
      try {
        if (typeof params?.title === "string") body.title = params.title.trim();
        if (typeof params?.disclosure === "string") body.disclosure = params.disclosure;

        if (typeof params?.uri === "string" && params.uri.trim()) {
          const target = resolveMemoryLocator(params, {
            defaultDomain: pluginCfg.defaultDomain,
            domainKey: "domain",
            pathKey: "parent_path",
            uriKey: "uri",
            allowEmptyPath: false,
            label: "uri",
          });
          const derived = splitParentPathAndTitle(target.path);
          if (!derived.title) {
            throw new Error("Create target URI must include a final path segment, like project://workflow/browser_policy");
          }
          if (typeof params?.title === "string" && params.title.trim() && params.title.trim() !== derived.title) {
            throw new Error(`Conflicting uri and title: ${derived.title} vs ${params.title.trim()}`);
          }
          body.domain = target.domain;
          body.parent_path = derived.parentPath;
          body.title = derived.title;
        }

        const data = await fetchJson(pluginCfg, `/browse/node`, { method: "POST", body: JSON.stringify(body) });
        const nodeUuid = String(data?.node_uuid || "").trim();
        const glossaryResult = nodeUuid && glossary.length > 0
          ? await applyGlossaryMutations(pluginCfg, nodeUuid, { add: glossary })
          : { added: [], removed: [] };
        const suffix = glossaryResult.added.length > 0 ? `\nGlossary: ${glossaryResult.added.join(", ")}` : "";
        return textResult(`Created ${data?.uri || `${body.domain}://${body.parent_path}`}${suffix}`, { ok: true, result: data, glossary: glossaryResult });
      } catch (error: any) {
        return textResult(`Lore create failed: ${error.message}`, { ok: false, error: error.message, body, glossary });
      }
    },
  });

  api.registerTool({
    name: "lore_update_node",
    label: "Lore update node",
    description: "Revise an existing long-term memory node when stored knowledge becomes clearer, newer, or more accurate.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["uri"],
      properties: {
        uri: { type: "string", description: "Full memory URI for the node you want to revise." },
        content: { type: "string" },
        priority: { type: "integer", minimum: 0 },
        disclosure: { type: "string" },
        glossary_add: { type: "array", items: { type: "string" } },
        glossary_remove: { type: "array", items: { type: "string" } }
      }
    },
    async execute(_id: any, params: any) {
      const body: any = {};
      const glossaryAdd = normalizeKeywordList(params?.glossary_add);
      const glossaryRemove = normalizeKeywordList(params?.glossary_remove);
      if (typeof params?.content === "string") body.content = params.content;
      if (Number.isFinite(params?.priority)) body.priority = params.priority;
      if (typeof params?.disclosure === "string") body.disclosure = params.disclosure;
      let domain = pluginCfg.defaultDomain;
      let path = "";
      try {
        ({ domain, path } = resolveMemoryLocator(params, { defaultDomain: pluginCfg.defaultDomain, pathKey: "__unused_path", allowEmptyPath: false, label: "uri" }));
        const qs = new URLSearchParams({ domain, path });
        const data = await fetchJson(pluginCfg, `/browse/node?${qs.toString()}`, { method: "PUT", body: JSON.stringify(body) });
        let glossaryResult: { added: string[]; removed: string[] } = { added: [], removed: [] };
        if (glossaryAdd.length > 0 || glossaryRemove.length > 0) {
          const nodeData = await fetchJson(pluginCfg, `/browse/node?${qs.toString()}`, { method: "GET" });
          const nodeUuid = String(nodeData?.node?.node_uuid || "").trim();
          if (!nodeUuid) throw new Error(`Node UUID not found for ${domain}://${path}`);
          glossaryResult = await applyGlossaryMutations(pluginCfg, nodeUuid, { add: glossaryAdd, remove: glossaryRemove });
        }
        const suffixParts: string[] = [];
        if (glossaryResult.added.length > 0) suffixParts.push(`glossary+ ${glossaryResult.added.join(", ")}`);
        if (glossaryResult.removed.length > 0) suffixParts.push(`glossary- ${glossaryResult.removed.join(", ")}`);
        const suffix = suffixParts.length > 0 ? `\n${suffixParts.join("\n")}` : "";
        return textResult(`Updated ${domain}://${path}${suffix}`, { ok: true, result: data, glossary: glossaryResult });
      } catch (error: any) {
        return textResult(`Lore update failed: ${error.message}`, { ok: false, error: error.message, domain, path, glossary_add: glossaryAdd, glossary_remove: glossaryRemove });
      }
    },
  });

  api.registerTool({
    name: "lore_delete_node",
    label: "Lore delete node",
    description: "Remove a memory path that is obsolete, duplicated, or no longer wanted.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["uri"],
      properties: {
        uri: { type: "string", description: "Full memory URI for the path you want to remove." }
      }
    },
    async execute(_id: any, params: any) {
      let domain = pluginCfg.defaultDomain;
      let path = "";
      try {
        ({ domain, path } = resolveMemoryLocator(params, { defaultDomain: pluginCfg.defaultDomain, pathKey: "__unused_path", allowEmptyPath: false, label: "uri" }));
        const qs = new URLSearchParams({ domain, path });
        const data = await fetchJson(pluginCfg, `/browse/node?${qs.toString()}`, { method: "DELETE" });
        return textResult(`Deleted ${domain}://${path}`, { ok: true, result: data });
      } catch (error: any) {
        return textResult(`Lore delete failed: ${error.message}`, { ok: false, error: error.message, domain, path });
      }
    },
  });

  api.registerTool({
    name: "lore_move_node",
    label: "Lore move node",
    description: "Move or rename a memory node to a new URI path. Updates all child paths automatically.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["old_uri", "new_uri"],
      properties: {
        old_uri: { type: "string", description: "Current memory URI to move from." },
        new_uri: { type: "string", description: "New memory URI to move to." },
      }
    },
    async execute(_id: any, params: any) {
      const body = {
        old_uri: String(params?.old_uri || "").trim(),
        new_uri: String(params?.new_uri || "").trim(),
      };
      try {
        const data = await fetchJson(pluginCfg, `/browse/move`, { method: "POST", body: JSON.stringify(body) });
        return textResult(`Moved ${body.old_uri} → ${body.new_uri}`, { ok: true, result: data });
      } catch (error: any) {
        return textResult(`Lore move failed: ${error.message}`, { ok: false, error: error.message, body });
      }
    },
  });

  api.registerTool({
    name: "lore_list_session_reads",
    label: "Lore list session reads",
    description: "Show which memory nodes have already been opened in this session.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["session_id"],
      properties: { session_id: { type: "string" } }
    },
    async execute(_id: any, params: any) {
      const sessionId = String(params?.session_id || "").trim();
      try {
        const qs = new URLSearchParams({ session_id: sessionId });
        const data = await fetchJson(pluginCfg, `/browse/session/read?${qs.toString()}`, { method: "GET" });
        const text = Array.isArray(data) && data.length > 0
          ? data.map((item: any) => `- ${item.uri} (${item.read_count})`).join("\n")
          : "No read nodes tracked for this session.";
        return textResult(text, { ok: true, reads: data });
      } catch (error: any) {
        return textResult(`Lore session reads failed: ${error.message}`, { ok: false, error: error.message });
      }
    },
  });

  api.registerTool({
    name: "lore_clear_session_reads",
    label: "Lore clear session reads",
    description: "Reset per-session memory read tracking.",
    parameters: {
      type: "object",
      additionalProperties: false,
      required: ["session_id"],
      properties: { session_id: { type: "string" } }
    },
    async execute(_id: any, params: any) {
      const sessionId = String(params?.session_id || "").trim();
      try {
        const qs = new URLSearchParams({ session_id: sessionId });
        const data = await fetchJson(pluginCfg, `/browse/session/read?${qs.toString()}`, { method: "DELETE" });
        return textResult(`Cleared Lore read tracking for ${sessionId}`, { ok: true, result: data });
      } catch (error: any) {
        return textResult(`Lore clear session reads failed: ${error.message}`, { ok: false, error: error.message });
      }
    },
  });
}

File: hermes-plugin/lore_hermes/tools.py
"""
Tools - Hermes tool registration for Lore

This module provides tools that can be registered with Hermes Agent
to interact with the Lore memory system.
"""

from typing import Dict, List, Optional, Any, Callable
from .client import LoreClient, LoreError
from .formatters import (
    format_node,
    format_boot_view,
    format_search_results,
    format_domains,
    format_session_reads,
    format_recall_block
)


class ToolResult:
    """Tool execution result"""
    def __init__(self, text: str, ok: bool = True, **kwargs):
        self.text = text
        self.ok = ok
        self.details = kwargs
    
    def __str__(self) -> str:
        return self.text


def register_tools(register_fn: Callable[[str, str, Dict, Callable], None], client: Optional[LoreClient] = None):
    """
    Register all Lore tools with Hermes
    
    Args:
        register_fn: Function to register a tool (name, description, schema, handler)
        client: Optional LoreClient instance (creates default if not provided)
    """
    if client is None:
        client = LoreClient()
    
    # ---- Status & Boot ----
    
    def lore_status() -> ToolResult:
        """Check Lore server health"""
        try:
            data = client.health()
            return ToolResult(
                f"Lore online\n\n{str(data)}",
                ok=True,
                health=data,
                base_url=client.base_url
            )
        except LoreError as e:
            return ToolResult(
                f"Lore offline: {e}",
                ok=False,
                error=str(e),
                base_url=client.base_url
            )
    
    register_fn(
        "lore_status",
        "Check Lore memory backend availability and connection health",
        {},
        lore_status
    )
    
    def lore_boot() -> ToolResult:
        """Load boot memory view"""
        try:
            data = client.boot()
            content = format_boot_view(data)
            return ToolResult(content, ok=True, boot=data)
        except LoreError as e:
            return ToolResult(f"Lore boot failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_boot",
        "Load the boot memory view that restores long-term identity and core operating context",
        {},
        lore_boot
    )
    
    # ---- Node Operations ----
    
    def lore_get_node(uri: str, nav_only: bool = False, session_id: Optional[str] = None) -> ToolResult:
        """Get a memory node by URI"""
        try:
            domain, path = client.parse_uri(uri)
            data = client.get_node(domain, path, nav_only)
            
            node = data.get("node", {})
            if session_id and node.get("uri"):
                try:
                    client.mark_session_read(
                        session_id=session_id,
                        uri=node["uri"],
                        node_uuid=node.get("node_uuid"),
                        source="tool:lore_get_node"
                    )
                except:
                    pass  # Best effort
            
            return ToolResult(format_node(data), ok=True, node=node, children=data.get("children", []))
        except LoreError as e:
            return ToolResult(f"Lore get node failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_get_node",
        "Open a memory node to inspect its full content, metadata, and nearby structure",
        {
            "uri": {"type": "string", "description": "Full memory URI (e.g., core://soul)"},
            "nav_only": {"type": "boolean", "description": "Skip expensive glossary processing"},
            "session_id": {"type": "string", "description": "Session ID for read tracking"}
        },
        lore_get_node
    )
    
    def lore_create_node(
        content: str,
        priority: int,
        glossary: List[str],
        uri: Optional[str] = None,
        domain: Optional[str] = None,
        parent_path: Optional[str] = None,
        title: Optional[str] = None,
        disclosure: Optional[str] = None
    ) -> ToolResult:
        """Create a new memory node"""
        try:
            # Determine target location
            if uri:
                target_domain, target_path = client.parse_uri(uri)
                parts = target_path.split("/")
                if not title and parts:
                    title = parts[-1]
                    parent_path = "/".join(parts[:-1]) if len(parts) > 1 else ""
                else:
                    parent_path = target_path
                domain = target_domain
            else:
                domain = domain or client.default_domain
                parent_path = parent_path or ""
            
            data = client.create_node(
                domain=domain,
                parent_path=parent_path,
                content=content,
                priority=priority,
                title=title,
                disclosure=disclosure
            )
            
            # Add glossary keywords
            node_uuid = data.get("node_uuid")
            added_glossary = []
            if node_uuid and glossary:
                for keyword in glossary:
                    try:
                        client.add_glossary(keyword, node_uuid)
                        added_glossary.append(keyword)
                    except:
                        pass
            
            suffix = f"\nGlossary: {', '.join(added_glossary)}" if added_glossary else ""
            return ToolResult(
                f"Created {data.get('uri', f'{domain}://{parent_path}')}{suffix}",
                ok=True,
                result=data,
                glossary=added_glossary
            )
        except LoreError as e:
            return ToolResult(f"Lore create failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_create_node",
        "Create a new long-term memory node for durable facts, rules, project knowledge",
        {
            "content": {"type": "string", "description": "Memory content"},
            "priority": {"type": "integer", "description": "Importance tier (0=core, 1=key, 2+=general)"},
            "glossary": {"type": "array", "items": {"type": "string"}, "description": "Keywords for search indexing"},
            "uri": {"type": "string", "description": "Full target URI (optional)"},
            "domain": {"type": "string", "description": "Target domain (if not using uri)"},
            "parent_path": {"type": "string", "description": "Parent path (if not using uri)"},
            "title": {"type": "string", "description": "Final path segment"},
            "disclosure": {"type": "string", "description": "When to recall this memory"}
        },
        lore_create_node
    )
    
    def lore_update_node(
        uri: str,
        content: Optional[str] = None,
        priority: Optional[int] = None,
        disclosure: Optional[str] = None,
        glossary_add: Optional[List[str]] = None,
        glossary_remove: Optional[List[str]] = None
    ) -> ToolResult:
        """Update an existing memory node"""
        try:
            domain, path = client.parse_uri(uri)
            
            # Update node
            data = client.update_node(
                domain=domain,
                path=path,
                content=content,
                priority=priority,
                disclosure=disclosure
            )
            
            # Handle glossary mutations
            added = []
            removed = []
            if glossary_add or glossary_remove:
                # Get node UUID
                node_data = client.get_node(domain, path)
                node_uuid = node_data.get("node", {}).get("node_uuid")
                
                if node_uuid:
                    for keyword in glossary_add or []:
                        try:
                            client.add_glossary(keyword, node_uuid)
                            added.append(keyword)
                        except:
                            pass
                    for keyword in glossary_remove or []:
                        try:
                            client.remove_glossary(keyword, node_uuid)
                            removed.append(keyword)
                        except:
                            pass
            
            parts = [f"Updated {domain}://{path}"]
            if added:
                parts.append(f"glossary+ {', '.join(added)}")
            if removed:
                parts.append(f"glossary- {', '.join(removed)}")
            
            return ToolResult("\n".join(parts), ok=True, result=data)
        except LoreError as e:
            return ToolResult(f"Lore update failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_update_node",
        "Revise an existing long-term memory node",
        {
            "uri": {"type": "string", "description": "Full memory URI"},
            "content": {"type": "string", "description": "New content"},
            "priority": {"type": "integer", "description": "New priority"},
            "disclosure": {"type": "string", "description": "New disclosure"},
            "glossary_add": {"type": "array", "items": {"type": "string"}},
            "glossary_remove": {"type": "array", "items": {"type": "string"}}
        },
        lore_update_node
    )
    
    def lore_delete_node(uri: str) -> ToolResult:
        """Delete a memory node"""
        try:
            domain, path = client.parse_uri(uri)
            data = client.delete_node(domain, path)
            return ToolResult(f"Deleted {domain}://{path}", ok=True, result=data)
        except LoreError as e:
            return ToolResult(f"Lore delete failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_delete_node",
        "Remove a memory path that is obsolete, duplicated, or no longer wanted",
        {"uri": {"type": "string", "description": "Full memory URI to delete"}},
        lore_delete_node
    )
    
    def lore_move_node(old_uri: str, new_uri: str) -> ToolResult:
        """Move/rename a memory node"""
        try:
            data = client.move_node(old_uri, new_uri)
            return ToolResult(f"Moved {old_uri} → {new_uri}", ok=True, result=data)
        except LoreError as e:
            return ToolResult(f"Lore move failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_move_node",
        "Move or rename a memory node to a new URI path",
        {
            "old_uri": {"type": "string", "description": "Current URI"},
            "new_uri": {"type": "string", "description": "New URI"}
        },
        lore_move_node
    )
    
    # ---- Search & Recall ----
    
    def lore_search(query: str, domain: Optional[str] = None, limit: int = 10) -> ToolResult:
        """Search memories by keyword"""
        try:
            data = client.search(query, domain=domain, limit=limit)
            results = data.get("results", [])
            meta = data.get("meta")
            text = format_search_results(results, meta)
            return ToolResult(text, ok=True, results=results, meta=meta)
        except LoreError as e:
            return ToolResult(f"Lore search failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_search",
        "Find relevant memories by keyword or domain",
        {
            "query": {"type": "string", "description": "Search query"},
            "domain": {"type": "string", "description": "Optional domain filter"},
            "limit": {"type": "integer", "description": "Max results (1-100)", "default": 10}
        },
        lore_search
    )
    
    # ---- Domains ----
    
    def lore_list_domains() -> ToolResult:
        """List all memory domains"""
        try:
            domains = client.list_domains()
            text = format_domains(domains)
            return ToolResult(text, ok=True, domains=domains)
        except LoreError as e:
            return ToolResult(f"Lore list domains failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_list_domains",
        "Browse the top-level memory domains available",
        {},
        lore_list_domains
    )
    
    # ---- Session Tracking ----
    
    def lore_list_session_reads(session_id: str) -> ToolResult:
        """List session read tracking"""
        try:
            reads = client.list_session_reads(session_id)
            text = format_session_reads(reads)
            return ToolResult(text, ok=True, reads=reads)
        except LoreError as e:
            return ToolResult(f"Lore session reads failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_list_session_reads",
        "Show which memory nodes have already been opened in this session",
        {"session_id": {"type": "string", "description": "Session ID"}},
        lore_list_session_reads
    )
    
    def lore_clear_session_reads(session_id: str) -> ToolResult:
        """Clear session read tracking"""
        try:
            data = client.clear_session_reads(session_id)
            return ToolResult(f"Cleared Lore read tracking for {session_id}", ok=True, result=data)
        except LoreError as e:
            return ToolResult(f"Lore clear session reads failed: {e}", ok=False, error=str(e))
    
    register_fn(
        "lore_clear_session_reads",
        "Reset per-session memory read tracking",
        {"session_id": {"type": "string", "description": "Session ID"}},
        lore_clear_session_reads
    )
